File: src/sc_neurocore/accel/julia/quantum/noise_models.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia acceleration for quantum/noise_models

module NoiseModelsAccel

using Statistics, LinearAlgebra

mutable struct HeronR2NoiseModelState
    cx_error::Float64
    single_qubit_error::Float64
    t1_us::Float64
    t2_us::Float64
    readout_0to1::Float64
    readout_1to0::Float64
    gate_time_1q_ns::Float64
    gate_time_2q_ns::Float64
end

function HeronR2NoiseModelState()
    HeronR2NoiseModelState(0.005, 0.0003, 300.0, 200.0, 0.01, 0.02, 25.0, 100.0)
end

function depolarizing_channel(s::HeronR2NoiseModelState, p)
    I = np.eye(2, dtype=complex)
    X = collect([[0, 1], [1, 0]], dtype=complex)
    Y = collect([[0, -1j], [1j, 0]], dtype=complex)
    Z = collect([[1, 0], [0, -1]], dtype=complex)
    return [
        sqrt(1 - p) * I,
        sqrt(p / 3) * X,
        sqrt(p / 3) * Y,
        sqrt(p / 3) * Z,
    ]
end

function amplitude_damping(s::HeronR2NoiseModelState, gamma)
    K0 = collect([[1, 0], [0, sqrt(1 - gamma)]], dtype=complex)
    K1 = collect([[0, sqrt(gamma)], [0, 0]], dtype=complex)
    return [K0, K1]
end

function phase_damping(s::HeronR2NoiseModelState, gamma)
    K0 = collect([[1, 0], [0, sqrt(1 - gamma)]], dtype=complex)
    K1 = collect([[0, 0], [0, sqrt(gamma)]], dtype=complex)
    return [K0, K1]
end

function apply_single_qubit_noise(s::HeronR2NoiseModelState, rho, Any])
    kraus = s.depolarizing_channel(s.params.single_qubit_error)
    return sum(K @ rho @ K.conj().T for K in kraus)
end

function apply_readout_noise(s::HeronR2NoiseModelState, measurement)
    p = s.params
    if measurement == 0
        return 1 if np.random.random() < p.readout_0to1 else 0
    return 0 if np.random.random() < p.readout_1to0 else 1
end

function gate_fidelity_1q(s::HeronR2NoiseModelState)
    return 1.0 - s.params.single_qubit_error
end

function gate_fidelity_2q(s::HeronR2NoiseModelState)
    return 1.0 - s.params.cx_error
end

end # module NoiseModelsAccel

File: src/sc_neurocore/accel/rust/safety/noise_models.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Rust safety for noise_models

#![allow(unused_variables, dead_code, non_snake_case)]

#[derive(Debug, Clone)]
pub struct HeronR2NoiseModel {
    pub cx_error: f64,
    pub single_qubit_error: f64,
    pub t1_us: f64,
    pub t2_us: f64,
    pub readout_0to1: f64,
    pub readout_1to0: f64,
    pub gate_time_1q_ns: f64,
    pub gate_time_2q_ns: f64,
}

impl HeronR2NoiseModel {
    pub fn new() -> Self {
        Self {
            cx_error: 0.005_f64,
            single_qubit_error: 0.0003_f64,
            t1_us: 300.0_f64,
            t2_us: 200.0_f64,
            readout_0to1: 0.01_f64,
            readout_1to0: 0.02_f64,
            gate_time_1q_ns: 25.0_f64,
            gate_time_2q_ns: 100.0_f64,
        }
    }

    pub fn depolarizing_channel(&self, p: f64) -> f64 {
        // I = np.eye(2, dtype=complex)
        // X = np.array([[0, 1], [1, 0]], dtype=complex)
        // Y = np.array([[0, -1j], [1j, 0]], dtype=complex)
        // Z = np.array([[1, 0], [0, -1]], dtype=complex)
        // return [
        // (1 - p_f64).sqrt() * I,
        // (p / 3_f64).sqrt() * X,
        // (p / 3_f64).sqrt() * Y,
        // (p / 3_f64).sqrt() * Z,
        // ]
        0.0
    }

    pub fn amplitude_damping(&self, gamma: f64) -> f64 {
        // K0 = np.array([[1, 0], [0, (1 - gamma_f64).sqrt()]], dtype=complex)
        // K1 = np.array([[0, (gamma_f64).sqrt()], [0, 0]], dtype=complex)
        // return [K0, K1]
        0.0
    }

    pub fn phase_damping(&self, gamma: f64) -> f64 {
        // K0 = np.array([[1, 0], [0, (1 - gamma_f64).sqrt()]], dtype=complex)
        // K1 = np.array([[0, 0], [0, (gamma_f64).sqrt()]], dtype=complex)
        // return [K0, K1]
        0.0
    }

    pub fn apply_single_qubit_noise(&self, rho: f64) -> f64 {
        // kraus = self.depolarizing_channel(self.params.single_qubit_error)
        // return sum(K @ rho @ K.conj().T for K in kraus)
        0.0
    }

    pub fn apply_readout_noise(&self, measurement: f64) -> f64 {
        // p = self.params
        // if measurement == 0:
        // return 1 if np.random.random() < p.readout_0to1 else 0
        // return 0 if np.random.random() < p.readout_1to0 else 1
        0.0
    }

    pub fn gate_fidelity_1q(&self, ) -> f64 {
        // return 1.0 - self.params.single_qubit_error
        0.0
    }

    pub fn gate_fidelity_2q(&self, ) -> f64 {
        // return 1.0 - self.params.cx_error
        0.0
    }

}

pub fn validate_noise_models(state: &HeronR2NoiseModel) -> bool {
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_noise_models_new() {
        let state = HeronR2NoiseModel::new();
        assert!(validate_noise_models(&state));
    }

}
